compiler/hash-typecheck/src/ops/simplify.rs

```diff
@@ -953,6 +974,85 @@ impl<'gs, 'ls, 'cd, 's> Simplifier<'gs, 'ls, 'cd, 's> {
         }
     }
 
+    /// Simplify the given set of terms by performing the following operations
+    /// (where ^ is the separator of the list):
+    ///
+    /// - Applying idempotency: B ^ B ^ C becomes B ^ C
+    /// - Flattening nests: B ^ (C ^ D) becomes B ^ C ^ D
+    /// - Simplifying inner terms:
+    ///  (<T> => (str, T))<i32> ^ C becomes (str, i32) ^ C
+    /// - Distributivity: @@Todo, @@Enhancement
+    ///
+    /// This is to be used for merge and union types.
+    pub fn simplify_algebraic_term_list(
+        &mut self,
+        term_list: &[TermId],
+        is_nested: impl Fn(&Term) -> Option<Vec<TermId>>,
+    ) -> TcResult<Option<Vec<TermId>>> {
+        let mut simplified_once = false;
+
+        // Flatten nests (associativity);
+        // Also simplify inner terms
+        let flattened = term_list
+            .iter()
+            .copied()
+            .map(|term_id| {
+                // Check if the term is a nested list, and if so flatten it:
+                let simplified_term_id = self
+                    .simplifier()
+                    .simplify_term(term_id)?
+                    .map(|x| {
+                        simplified_once = true;
+                        x
+                    })
+                    .unwrap_or(term_id);
+                let reader = self.reader();
+                let term = reader.get_term(simplified_term_id);
+                match is_nested(term) {
+                    // It is a merge, flatten it (this also means the merge has been
+                    // simplified):
+                    Some(terms) => {
+                        simplified_once = true;
+                        Ok(terms)
+                    }
+                    // Not a merge, just return a single-element vector:
+                    _ => Ok(vec![simplified_term_id]),
+                }
+            })
+            .try_fold(vec![], |mut all_terms, nested_terms| {
+                // Combine all the nested terms
+                all_terms.extend(nested_terms?);
+                Ok(all_terms)
+            })?;
+
+        // Merge equal terms (idempotency)
+        let mut merged: Vec<_> = flattened.into_iter().map(Some).collect();
+        for terms in term_list.iter().enumerate().combinations(2) {
+            match terms.as_slice() {
+                [(first_idx, &first), (second_idx, &second)] => {
+                    // Try to merge the two terms if they are the same:
+                    if self.unifier().terms_are_equal(first, second) {
+                        simplified_once = true;
+                        merged[*first_idx] = merged[*first_idx].map(|_| first);
+                        merged[*second_idx] = None;
+                    } else {
+                        merged[*first_idx] = merged[*first_idx].map(|_| first);
+                        merged[*second_idx] = merged[*second_idx].map(|_| second);
+                    }
+                }
+                _ => unreachable!(),
```

Comment:
maybe make this a `tc_panic`?

Reply:
Well this is not a bug in the typechecker if this happens, it is just an impossible situation because of `.combinations(2)` above. I think using `tc_panic` would be overkill.